src/spec-common/cliHost.ts

```diff
@@ -87,22 +87,126 @@ function createLocalCLIHostFromExecFunctions(localCwd: string, exec: ExecFunctio
+		const bytesInChar = number.match(/.{2}/g);
+		if (bytesInChar !== null) {
+			bytesInChar.reverse().map((byte) => {
+				bytes.push(parseInt(byte, 16));
+			});
+		}
+	});
+	return Buffer.from(bytes);
+}
+
+// The cygwin/git bash ssh-agent server will reply us with the cookie back (16 bytes)
+// + identifiers (12 bytes), skip them while forwarding data from ssh-agent to the client
+function skipHeader(headerSize: number, cb: SourceCallback<Buffer>, abort: Abort, data?: Buffer): number {
+	if (abort || data === undefined) {
+		cb(abort);
+		return headerSize;
 	}
 
-	const socket = new net.Socket();
+	if (headerSize === 0) {
+		// Fast path avoiding data buffer manipulation
+		// We don't need to modify the received data (handshake header
+		// already removed)
+		cb(null, data);
+	} else if (data.length > headerSize) {
+		// We need to remove part of the data to forward
+		data = data.slice(headerSize, data.length);
+		headerSize = 0;
+		cb(null, data);
+	} else {
+		// We need to remove all forwarded data
+		headerSize = headerSize - data.length;
+		cb(null, Buffer.of());
+	}
+
+	// Return the updated headerSize
+	return headerSize;
+}
+
+// Function to handle the Cygwin/Gpg4win socket filtering
+// These sockets need an handshake before forwarding client and server data
+function handleUnixSocketOnWindows(socket: net.Socket, socketPath: string): Duplex<Buffer, Buffer> {
+	let headerSize = 0;
+	let pendingSourceCallbacks: { abort: Abort; cb: SourceCallback<Buffer> }[] = [];
+	let pendingSinkCalls: Source<Buffer>[] = [];
+	let connectionDuplex: Duplex<Buffer, Buffer> | undefined = undefined;
+
 	(async () => {
 		const buf = await readLocalFile(socketPath);
-		const i = buf.indexOf(0xa);
-		const port = parseInt(buf.slice(0, i).toString(), 10);
-		const guid = buf.slice(i + 1);
+		const str = buf.toString();
+
+		// Try to parse cygwin socket data
+		const cygwinSocketParameters = str.match(/!<socket >(\d+)( s)? ((([A-Fa-f0-9]{2}){4}-?){4})/);
+
+		let port: number;
+		let handshake: Buffer;
+
+		if (cygwinSocketParameters !== null) {
+			// Cygwin / MSYS / Git Bash unix socket on Windows
+			const portStr = cygwinSocketParameters[1];
+			const guidStr = cygwinSocketParameters[3];
+			port = parseInt(portStr, 10);
+			const guid = cygwinUnixSocketCookieToBuffer(guidStr);
+
+			let identifierData = Buffer.alloc(12);
+			identifierData.writeUInt32LE(process.pid, 0);
+
+			handshake = Buffer.concat([guid, identifierData]);
+
+			// Recv header size = GUID (16 bytes) + identifiers (3 * 4 bytes)
+			headerSize = 16 + 3 * 4;
+		} else {
+			// Gpg4Win unix socket
+			const i = buf.indexOf(0xa);
+			port = parseInt(buf.slice(0, i).toString(), 10);
+			handshake = buf.slice(i + 1);
+
+			// No header will be received from Gpg4Win agent
+			headerSize = 0;
+		}
+
+		// Handle connection errors and resets
+		socket.on('error', err => {
+			console.error(err);
+			socket.destroy();
+
+			for (let callback of pendingSourceCallbacks) {
```

Comment:
Could these be old callbacks that were handled below already? E.g., when the error occurs after the handshake already finished. Maybe clear `pendingSourceCallbacks` after handling them (below and also here in case the error handler can be called a second time)?

Reply:
Indeed, will do that.

Should I also call `pendingSinkCalls` callbacks too ? I'm thinking about calling them like this: `callback(err, (_abort, _data) => { });`.
So if the caller doesn't do any read (doesn't call `source`) but only `sink`, it is also notified that the sink is closed.

Reply:
Makes sense, thanks!